vllm/model_executor/layers/sampler.py

```diff
@@ -37,14 +37,17 @@ def forward(
         hidden_states: torch.Tensor,
         sampling_metadata: SamplingMetadata,
         embedding_bias: Optional[torch.Tensor] = None,
-    ) -> SamplerOutput:
+    ) -> Optional[SamplerOutput]:
         # Get the hidden states that we use for sampling.
         hidden_states = _prune_hidden_states(hidden_states, sampling_metadata)
 
         # Get the logits for the next tokens.
         logits = _get_logits(hidden_states, embedding, embedding_bias,
                              self.vocab_size)
 
+        if not sampling_metadata.perform_sampling:
```

Comment:
Why not putting this at the beginning of this method?

Reply:
Is logits computation a process of sampling? It seems a middle state process between model and sampling. 🤔

Reply:
I kinda agree with @esmeetu. Maybe we should make it a separate stage. But let's discuss it in a future PR.

Reply:
Added a TODO here